<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
# Подключение к БД
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
    SELECT COUNT(*)
    FROM public.vacancies
'''

In [5]:
df = pd.read_sql_query(query_3_1, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2503887465.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''
    SELECT COUNT(*)
    FROM public.employers
'''

In [7]:
df = pd.read_sql_query(query_3_2, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\438664466.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''
    SELECT COUNT(*)
    FROM public.areas
'''

In [9]:
df = pd.read_sql_query(query_3_3, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2888627660.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''
    SELECT COUNT(*)
    FROM public.industries
'''

In [11]:
df = pd.read_sql_query(query_3_4, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\181467837.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

В базе данных храниться информация о 49197 вакансиях, также работодателей в базе данных 23501, что говорит о том что у одного работодателя может быть несколько вакансий на разные сферы деятельности и в разных регионах, даже если регион один то сферы деятельности в которые работодателю требуются сотрудники могут быть разные, или же в сфера деятельности может быть одна, но работодателю могут требоваться люди на разные позиции.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''
    SELECT a.name AS area, COUNT(v.name) AS cnt
    FROM public.vacancies AS v
    JOIN public.areas AS a ON v.area_id = a.id
    GROUP BY a.name
    ORDER BY COUNT(v.name) DESC
    LIMIT 5
'''

In [13]:
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2004364207.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''
    SELECT COUNT(name)
    FROM public.vacancies
    WHERE (salary_from IS NOT NULL) OR (salary_to IS NOT NULL)
'''

In [15]:
df = pd.read_sql_query(query_4_2, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\1373709454.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
query_4_3 = f'''
    SELECT AVG(salary_from) AS mean_salary_from, AVG(salary_to) AS mean_salary_to
    FROM public.vacancies
'''

In [18]:
df = pd.read_sql_query(query_4_3, connection)
display(df.round())

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\629610214.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


mean_salary_from  mean_salary_to
0           71065.0        110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
query_4_4 = f'''
    SELECT employment, schedule, COUNT(*)
    FROM public.vacancies
    GROUP BY employment, schedule
    ORDER BY COUNT(*) DESC
    LIMIT 2
'''

In [20]:
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\4268832789.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


employment          schedule  count
0  Полная занятость       Полный день  35367
1  Полная занятость  Удаленная работа   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
query_4_5 = f'''
    SELECT experience, COUNT(*)
    FROM public.vacancies
    GROUP BY experience
    ORDER BY COUNT(*) ASC
'''

In [22]:
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\3611053821.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Из данного анализа можно сделать вывод что чаще всего ищут людей либо с большим опытом работы или же тех у кого опыта работы нет, также многие люди ищут удаленную работу, это может быть связанно с тем что большинство вакансий находиться в таких крупных городах как Москва и Санкт-Петербург, и не все люди проживают на территории данных городов, также это можно объяснить тем что и зарплаты в данных городах выше чем по регионам и в среднем составляют от 71 000 рублей до 110 000 рублей

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
query_5_1 = f'''
    SELECT e.name AS name_employer, COUNT(*)
    FROM public.vacancies AS v
    JOIN public.employers AS e ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY COUNT(*) DESC
    LIMIT 5
'''

In [24]:
df = pd.read_sql_query(query_5_1, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2915140023.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name_employer  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
query_5_2 = f'''
    SELECT a.name, COUNT(distinct e.id) AS count_employer, COUNT(v.id) AS count_vacancies
    FROM public.areas AS a 
    JOIN public.employers AS e ON a.id = e.area
    LEFT JOIN public.vacancies AS v ON v.area_id = a.id
    GROUP BY a.id
    ORDER BY COUNT(v.id), COUNT(e.id) DESC
'''

In [26]:
df = pd.read_sql_query(query_5_2, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2644321718.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count_employer  count_vacancies
0                Россия             410                0
1             Казахстан             207                0
2    Московская область              75                0
3    Краснодарский край              19                0
4              Беларусь              18                0
..                  ...             ...              ...
649         Новосибирск             573          1149438
650              Алматы             721          1364132
651               Минск            1115          2354880
652     Санкт-Петербург            2217          6320667
653              Москва            5864         31272712

[654 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
query_5_3 = f'''
    SELECT e.name, COUNT(DISTINCT v.area_id) AS count_area
    FROM public.employers AS e
    JOIN public.vacancies AS v ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY COUNT(v.area_id) DESC
    LIMIT 1
'''

In [28]:
df = pd.read_sql_query(query_5_3, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\1682211407.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count_area
0  Яндекс         181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
query_5_4 = f'''
    SELECT COUNT(*)
    FROM public.employers AS e
    LEFT JOIN public.employers_industries AS ei ON e.id = ei.employer_id
    WHERE ei.employer_id IS NULL
'''

In [30]:
df = pd.read_sql_query(query_5_4, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\1196796479.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
query_5_5 = f'''
    SELECT e.name, COUNT(ei.industry_id)
    FROM public.employers AS e
    JOIN public.employers_industries AS ei ON e.id = ei.employer_id
    GROUP BY e.name
    HAVING COUNT(ei.industry_id) = 4
    ORDER BY e.name
    OFFSET 2 LIMIT 1
'''

In [32]:
df = pd.read_sql_query(query_5_5, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\1337360513.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [10]:
query_5_6 = f'''
    SELECT COUNT(DISTINCT e.id)
    FROM public.employers_industries AS ei
    JOIN public.employers AS e ON e.id = ei.employer_id
    JOIN public.industries AS i ON i.id = ei.industry_id
    WHERE i.name = 'Разработка программного обеспечения'
'''

In [11]:
df = pd.read_sql_query(query_5_6, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_1420\2317649464.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
# код для получения списка городов-милионников

In [36]:
# текст запроса

In [37]:
# результат запроса

***

Первый вывод который можно сделать опираясь на данные компания "Яндекс", является работодателем с большим количеством вакансий, это может быть связанно с тем что данная компания может привлекать людей с разных регионов на удаленной основе, а причины потребности в таком количестве кадров могут быть связанны, как и с ростом компании и распространением ее влияния, также и с внутренней "текучкой" кадров.

Также не все работодатели указывают сферу деятельности, это может быть связанно с тем что компании являются не большими и занимаются одной сферой деятельности, которая напрямую указывается через вакансию и обычно в основной информации компании, другими словами не у всех компаний есть потребность в указывании сферы деятельности.

Также есть компании с большим количеством сфер деятельности, это может говорить о том что компания является крупной, раз у нее есть охват нескольких сфер деятельности.

Также количество работодателей у которых в качестве сферы деятельности указана «Разработка программного обеспечения», составляет 3553 из общего количества 15082, что составляет 23% от количества всех сфер деятельности

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
query_6_1 = f'''
    SELECT COUNT(v.name)
    FROM public.vacancies AS v
    WHERE v.name ILIKE '%данн%' OR v.name ILIKE '%data%'
'''

In [40]:
df = pd.read_sql_query(query_6_1, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\3403335449.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
# текст запроса
query_6_2 = f'''
      SELECT COUNT(*)
      FROM public.vacancies AS v
      WHERE (v.name ILIKE '%data scientist%' 
            OR v.name ILIKE '%data science%'
            OR v.name ILIKE '%исследователь данных%'
            OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%')
            AND 
            (v.name ILIKE '%junior%' 
            OR v.experience ILIKE '%Нет опыта%'
            OR v.employment ILIKE '%Стажировка%')
'''

In [42]:
df = pd.read_sql_query(query_6_2, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\776391393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
query_6_3 = f'''
      SELECT COUNT(*)
      FROM public.vacancies AS v
      WHERE (v.name ILIKE '%data scientist%' 
            OR v.name ILIKE '%data science%'
            OR v.name ILIKE '%исследователь данных%'
            OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%' AND v.name NOT ILIKE '%TeamLead%')
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%')
            AND 
            (v.key_skills ILIKE '%SQL%' 
            OR v.key_skills ILIKE '%postgres%')
'''

In [44]:
df = pd.read_sql_query(query_6_3, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2765716463.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
query_6_4 = f'''
      SELECT COUNT(*)
      FROM public.vacancies AS v
      WHERE (v.name ILIKE '%data scientist%' 
            OR v.name ILIKE '%data science%'
            OR v.name ILIKE '%исследователь данных%'
            OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%' AND v.name NOT ILIKE '%TeamLead%')
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%')
            AND v.key_skills ILIKE '%Python%' 
'''

In [46]:
df = pd.read_sql_query(query_6_4, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\3377110858.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5 = f'''
SELECT ROUND(AVG(1 + length(key_skills) - length(Replace(key_skills, CHR(9), ''))), 2)
FROM public.vacancies AS v
WHERE (LOWER(v.name) LIKE '%data scientist%' 
      OR LOWER(v.name) LIKE '%data science%'
      OR LOWER(v.name) LIKE '%исследователь данных%'
      OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%')
      OR LOWER(v.name) LIKE '%machine learning%'
      OR LOWER(v.name) LIKE '%машинн%обучен%')
'''

In [48]:
df = pd.read_sql_query(query_6_5, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\2442488288.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# текст запроса
query_6_6 = f'''
SELECT ROUND(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0)
FROM public.vacancies AS v
WHERE (v.name ILIKE '%data scientist%' 
      OR v.name ILIKE '%data science%'
      OR v.name ILIKE '%исследователь данных%'
      OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%' AND v.name NOT ILIKE '%TeamLead%')
      OR v.name ILIKE '%machine learning%'
      OR v.name ILIKE '%машинн%обучен%') 
      AND v.experience ILIKE 'От 3 до 6 лет'
'''

In [50]:
df = pd.read_sql_query(query_6_6, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_7080\4002724519.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


round
0  243115.0

***

Вакансии связанных с DataSience занимают примерно 3.6% от общего числа вакансий, где 51 вакансия относиться к позиции Junior.

# Общий вывод по проекту

Каких вакансий много?

Как видно из запроса ниже больше всего вакансий связанных с ИТ-сферой

In [18]:
# текст запроса
count_vacancies = f'''
SELECT v.name, COUNT(*)
FROM public.vacancies AS v
GROUP BY v.name
ORDER BY COUNT(*) DESC
LIMIT 30
'''

df = pd.read_sql_query(count_vacancies, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_1420\2142893837.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(count_vacancies, connection)


name  count
0                             Системный администратор   1761
1                                      Программист 1С   1153
2                                     Бизнес-аналитик    488
3                    Специалист технической поддержки    457
4                                  Системный аналитик    432
5                                            Аналитик    404
6                                 Инженер-программист    389
7                                      Программист 1C    360
8                                Frontend-разработчик    242
9                       Инженер технической поддержки    230
10                                        Программист    224
11                               Руководитель проекта    191
12            Младший специалист ручного тестирования    169
13        Специалист службы поддержки в Яндекс Бизнес    168
14                                  Менеджер проектов    153
15                                   Java-разработчик    150
16          Специалист по информационной безопасности    148
17  Специалист службы поддержки Рекламной сети Янд...    139
18                                    PHP-разработчик    135
19          Специалист службы поддержки пользователей    126
20                                Компьютерный мастер    123
21                                   Менеджер проекта    120
22                               Технический писатель    111
23                                Маркетолог-аналитик    108
24                              Руководитель проектов    106
25                             Ведущий программист 1C    106
26                             Технический специалист    104
27                 Сотрудник поддержки поиска Яндекса     99
28                    Ведущий системный администратор     99
29         Технический писатель (дистанционный автор)     99

Что можно сказать о рынке труда в различных городах?

Из запроса ниже видно что более высокие зарплаты находятся в городах за пределами России

In [21]:
count_vacancies = f'''
    SELECT a.name, Round(Avg((Coalesce(salary_from, salary_to) + Coalesce(salary_to, salary_from)) / 2)) AS avg_salary
    FROM areas a
    JOIN vacancies v ON v.area_id = a.id
    WHERE (v.salary_to   IS NOT NULL 
    OR v.salary_from IS NOT NULL)
    GROUP BY a.name
    HAVING Count(v.id) > 10
    ORDER BY avg_salary DESC
    LIMIT 20
'''

df = pd.read_sql_query(count_vacancies, connection)
display(df)

C:\Users\Work_Machine_son\AppData\Local\Temp\ipykernel_1420\30171255.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(count_vacancies, connection)


name  avg_salary
0            Турция    235435.0
1              Кипр    223736.0
2           Армения    219570.0
3            Сербия    208423.0
4           Тбилиси    207156.0
5            Батуми    199834.0
6         Иннополис    175882.0
7              Баку    155747.0
8            Москва    134140.0
9             Лобня    121136.0
10          Магадан    117033.0
11       Зеленоград    116946.0
12      Новосибирск    107777.0
13  Санкт-Петербург    104410.0
14             Сочи     99437.0
15            Томск     99319.0
16            Химки     98641.0
17  Нижний Новгород     96059.0
18            Пермь     95985.0
19           Мытищи     94567.0